### PR TITLE
Colin/refactor

### LINKED ISSUE #4: AnteHandler
The ante handler is responsible for checking the following:
- inputs = outputs + fee
- Tx sig and confirm Sig are valid
- The Utxo's trying to be spent exist and have not been spent

The last check will be dependent upon our implementation of finalizing utxo positions.

### PR DESCRIPTION
**Pull request by colin-axner**
Closes #4 #10 #15 #16 #17 #18 

Still requires correct fee handling before merge to master. Ante.go has been updated so feeAmount is only updated and does not check for a feeUTXO. 

All tests currently pass.

### CODE REVIEW

File: auth/ante.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		return sdk.ErrUnknownRequest("UTXO trying to be spent, does not exist").Result()
+	}
+
+	//Checks that utxo owner equals address in the spendmsg
+	if !reflect.DeepEqual(utxo.GetAddress().Bytes(), addr.Bytes()) {
+		return sdk.ErrUnauthorized("signer does not match utxo owner").Result()
+	}
+
+	hash := ethcrypto.Keccak256(signBytes)
+	pubKey1, err1 := ethcrypto.SigToPub(hash, sig.Signature.Bytes()[5:])
```
Comment (legengliu) [inline auth/ante.go]: Why apply `[5:]` to the bytes?

Comment (colin-axner) [reply auth/ante.go]: Good question. 

The Bytes() function for crypto.SignatureSecp256K1 calls UnmarshalBinaryBare() as seen [here](https://github.com/tendermint/go-crypto/blob/master/signature.go#L64). 

When this occurs, the signature goes from 65 bytes to 70 bytes on my local test. The first five bytes are always the same regardless of the testing instance, on my local test they are [22 225 254 234 65]. According to the amino spec concrete structs should be prefixed with 4 bytes. I am still trying to figure out exactly why calling UnmarshalBinaryBare is adding those 5 bytes of padding to the front of the sig. I'd like to replace the "[5:]" with something more concrete based on the amino repo. 

Let me know if you figure anything out.

Comment (colin-axner) [reply auth/ante.go]: As per the amino spec, The first 4 bytes after the signature is marshaled corresponds to the prefix bytes, which are outlined [here](https://github.com/tendermint/tendermint/blob/26f633ed48441f72895b710f0e87b7b6c6791066/docs/specification/new-spec/encoding.md#signaturesecp256k1) . 

The 5th byte corresponds to the length of the byte array. Signatures should always be 65 bytes in length and unless the amino spec changes, there should continue to be 4 bytes used for prefixing. 

 `[5:]` could be replaced with `[amino.PrefixBytesLen + 1:]` but I think the former is sufficient for now, and the dependency on the amino spec should be noted in the code.

Comment (legengliu) [reply auth/ante.go]: I see. Good find, and yes agreed on including a comment in the code about the prefix bytes.

**Pull request closed without merging**